man/setmode.3
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)setmode.3	8.2 (Berkeley) 4/28/95
.\" $FreeBSD: src/lib/libc/gen/setmode.3,v 1.12 2007/01/09 00:27:55 imp Exp $
.\"
.Dd April 28, 1995
.Dt SETMODE 3
.Os
.Sh NAME
.Nm getmode ,
.Nm setmode
.Nd modify mode bits
.Sh LIBRARY
.ds str-Lb-libbsd Utility functions from BSD systems (libbsd, \-lbsd)
.Lb libbsd
.Sh SYNOPSIS
.In unistd.h
.Ft mode_t
.Fn getmode "const void *set" "mode_t mode"
.Ft void *
.Fn setmode "const char *mode_str"
.Sh DESCRIPTION
The
.Fn getmode
function
returns a copy of the file permission bits
.Fa mode
as altered by the values pointed to by
.Fa set .
While only the mode bits are altered, other parts of the file mode
may be examined.
.Pp
The
.Fn setmode
function
takes an absolute (octal) or symbolic value, as described in
.Xr chmod 1 ,
as an argument
and returns a pointer to mode values to be supplied to
.Fn getmode .
Because some of the symbolic values are relative to the file
creation mask,
.Fn setmode
may call
.Xr umask 2 .
If this occurs, the file creation mask will be restored before
.Fn setmode
returns.
If the calling program changes the value of its file creation mask
after calling
.Fn setmode ,
.Fn setmode
must be called again if
.Fn getmode
is to modify future file modes correctly.
.Pp
If the mode passed to
.Fn setmode
is invalid or if memory cannot be allocated for the return value,
.Fn setmode
returns
.Dv NULL .
.Pp
The value returned from
.Fn setmode
is obtained from
.Fn malloc
and should be returned to the system with
.Fn free
when the program is done with it, generally after a call to
.Fn getmode .
.Sh ERRORS
The
.Fn setmode
function
may fail and set errno for any of the errors specified for the library
routine
.Xr malloc 3 .
.Sh SEE ALSO
.Xr chmod 1 ,
.Xr stat 2 ,
.Xr umask 2 ,
.Xr malloc 3
.Sh HISTORY
The
.Fn getmode
and
.Fn setmode
functions first appeared in
.Bx 4.4 .


ChangeLog
2008-03-16  Guillem Jover  <guillem@debian.org>

	* include/vis.h: Remove UC Berkeley advertising clause, as per
	<ftp://ftp.cs.berkeley.edu/pub/4bsd/README.Impt.License.Change>.
	* src/vis.c: Likewise.
	* src/unvis.c: Likewise.
	* src/heapsort.c: Likewise.
	* src/linkaddr.c: Likewise.
	* man/fgetln.3: Likewise.

2007-12-18  Petr Salinger  <Petr.Salinger@seznam.cz>

	* include/bsd/queue.h: sync with current glibc variant on GNU/kFreeBSD

2006-11-08  Aurelien Jarno  <aurelien@aurel32.net>

	* src/link_addr.c: New file.
	* include/bsd/if_dl.h: Add link_addr.c declarations.
	* Makefile (LIB_SRCS): Add link_addr.c.
	* Versions: add link_addr and link_ntoa

2006-08-25  Petr Salinger  <Petr.Salinger@seznam.cz>

	* src/humanize_number.c: New file.
	* include/libutil.h: New file with humanize_number declarations.
	* Makefile (LIB_SRCS): Add humanize_number.c.
	* Versions: Add humanize_number.

2006-03-29  Robert Millan  <rmh@aybabtu.com>

	* src/heapsort.c: New file.
	* include/bsd/stdlib.h: Add heapsort declaration.
	* Makefile (LIB_SRCS): Add heapsort.c.
	* Versions: Add heapsort.

2006-03-19  Robert Millan  <rmh@aybabtu.com>

	* src/vis.c: New file.
	* src/unvis.c: Ditto.
	* include/vis.h: Add vis/unvis declarations.
	* Makefile (LIB_SRCS): Add vis.c/unvis.c.
	* Versions: Add vis/unvis functions.

2006-03-17  Robert Millan  <rmh@aybabtu.com>

	* src/progname.c: New file.
	* include/bsd/stdlib.h: Add {get,set}progname declarations.
	* Makefile (LIB_SRCS): Add 'progname.c'.
	* Versions: Add getprogname and setprogname.

2006-02-15  Guillem Jover  <guillem@debian.org>

	* include/bsd/cdefs.h: Disable __unused, it conflicts with a struct
	member in a Linux header.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/stdlib.h: Remove <sys/types.h> and <stdio.h>.
	* include/bsd/getopt.h: Protect header from multiple inclusion.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h: Remove commented out macros.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h (setproctitle, __dead2, __unused, __printflike)
	(__FBSDID): Moved to ...
	* include/bsd/cdefs.h: ... here.
	(__unused, __printflike) [__GNUC__]: Use proper __attribute__.
	* Makefile (LIB_INCLUDES): Add 'cdefs.h'.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* arc4random.c: Move to ...
	* src/arc4random.c: ... here.
	* bsd_getopt.c: Move to ...
	* src/bsd_getopt.c:: ... here.
	* err.c: Move to ...
	* src/err.c:: ... here.
	* fgetln.c: Move to ...
	* src/fgetln.c:: ... here.
	* fmtcheck.c: Move to ...
	* src/fmtcheck.c:: ... here.
	* inet_net_pton.c: Move to ...
	* src/inet_net_pton.c:: ... here.
	* strlcat.c: Move to ...
	* src/strlcat.c:: ... here.
	* strlcpy.c: Move to ...
	* src/strlcpy.c:: ... here.
	* md5c.c: Move to ...
	* src/md5c.c:: ... here.
	* md5.copyright: Move to ...
	* src/md5.copyright:: ... here.
	* Makefile (LIB_SRCS): Automatically prefix the src/ dir.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_SRCS): Line wrap.
	(LIB_INCLUDES): List only the files, automatically prefix the dir.
	(LIB_MANS): Likewise.

2006-02-13  Robert Millan  <rmh@aybabtu.com>

	* bsd_getopt.c: New file.
	* include/bsd/getopt.h: Likewise.
	* Makefile (LIB_SRCS): Add 'bsd_getopt.c'.
	(LIB_INCLUDES): Add 'include/bsd/getopt.h'.
	* Versions: Added bsd_getopt and optreset.

2006-02-10  Robert Millan  <rmh@aybabtu.com>

	* err.c: New file.
	* include/bsd/err.h: Likewise.
	* Makefile (LIB_SRCS): Add 'err.c'.
	(LIB_INCLUDES): Add 'include/bsd/err.h'.
	* Versions: Added errc, warnc, verrc and vwarn.

2005-12-19  Aurelien Jarno <aurel32@debian.org>

	* man/arc4random.3: New file.
	* man/fgetln.3: Likewise.
	* man/fmtcheck.3: Likewise.
	* man/strlcpy.3: Likewise.
	* fmtcheck.c:: Likewise.
	* include/stdlib.h: Added fmtcheck declaration.
	* Versions: Added fmtcheck.

2005-11-01  Guillem Jover  <guillem@debian.org>

	* ip_icmp.h: Move to ...
	* include/bsd/ip_icmp.h: ... here.
	* md5.h: Move to ...
	* include/bsd/md5.h: ... here.
	* bsd.h: Move to ...
	* include/bsd/bsd.h: ... here. Split contents ...
	* include/bsd/random.h: ... here ...
	* include/bsd/queue.h: ... here ...
	* include/bsd/string.h: ... and here.
	* md5c.c: Change md5.h include line.
	* Makefile (MK_CFLAGS): Include from the proper dir.

2005-08-03  Hector Garcia  <hector@debian.org>

	* Makefile: Fixed to place soft links instead in soname, and
	shared_so instead of copying files.
	* md5.h (MD5End, MD5File, MD5FileChunk, MD5Data): Commented out
	functions not available on 'md5c.c'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_INCLUDES): Add 'md5.h'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Versions (LIBBSD_0.0): Add MD5Init, MD5Update, MD5Pad and
	MD5Final.

2005-07-26  Hector Garcia  <hector@debian.org>

	* Makefile (LIB_SRCS): Add 'md5c.c'.
	* md5c.c: Changed include to 'md5.h'.
	* md5.h: Replaced with the system header 'sys/md5.h'.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* fgetln.c (fgetln): Reindent, remove commented code and translate
	variable names to english.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* Versions: New file.
	* Makefile: Add versioned symbols support.



src/setmode.c
/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Dave Borman at Cray Research, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)setmode.c	8.2 (Berkeley) 3/25/94";
#endif /* LIBC_SCCS and not lint */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/lib/libc/gen/setmode.c,v 1.11 2007/01/09 00:27:55 imp Exp $");

#include <sys/types.h>
#include <sys/stat.h>

#include <ctype.h>
#include <signal.h>
#include <stddef.h>
#include <stdlib.h>
#include <unistd.h>

#ifdef SETMODE_DEBUG
#include <stdio.h>
#endif

#define	SET_LEN	6		/* initial # of bitcmd struct to malloc */
#define	SET_LEN_INCR 4		/* # of bitcmd structs to add as needed */

typedef struct bitcmd {
	char	cmd;
	char	cmd2;
	mode_t	bits;
} BITCMD;

#define	CMD2_CLR	0x01
#define	CMD2_SET	0x02
#define	CMD2_GBITS	0x04
#define	CMD2_OBITS	0x08
#define	CMD2_UBITS	0x10

static BITCMD	*addcmd(BITCMD *, int, int, int, u_int);
static void	 compress_mode(BITCMD *);
#ifdef SETMODE_DEBUG
static void	 dumpmode(BITCMD *);
#endif

/*
 * Given the old mode and an array of bitcmd structures, apply the operations
 * described in the bitcmd structures to the old mode, and return the new mode.
 * Note that there is no '=' command; a strict assignment is just a '-' (clear
 * bits) followed by a '+' (set bits).
 */
mode_t
getmode(const void *bbox, mode_t omode)
{
	const BITCMD *set;
	mode_t clrval, newmode, value;

	set = (const BITCMD *)bbox;
	newmode = omode;
	for (value = 0;; set++)
		switch(set->cmd) {
		/*
		 * When copying the user, group or other bits around, we "know"
		 * where the bits are in the mode so that we can do shifts to
		 * copy them around.  If we don't use shifts, it gets real
		 * grundgy with lots of single bit checks and bit sets.
		 */
		case 'u':
			value = (newmode & S_IRWXU) >> 6;
			goto common;

		case 'g':
			value = (newmode & S_IRWXG) >> 3;
			goto common;

		case 'o':
			value = newmode & S_IRWXO;
common:			if (set->cmd2 & CMD2_CLR) {
				clrval =
				    (set->cmd2 & CMD2_SET) ?  S_IRWXO : value;
				if (set->cmd2 & CMD2_UBITS)
					newmode &= ~((clrval<<6) & set->bits);
				if (set->cmd2 & CMD2_GBITS)
					newmode &= ~((clrval<<3) & set->bits);
				if (set->cmd2 & CMD2_OBITS)
					newmode &= ~(clrval & set->bits);
			}
			if (set->cmd2 & CMD2_SET) {
				if (set->cmd2 & CMD2_UBITS)
					newmode |= (value<<6) & set->bits;
				if (set->cmd2 & CMD2_GBITS)
					newmode |= (value<<3) & set->bits;
				if (set->cmd2 & CMD2_OBITS)
					newmode |= value & set->bits;
			}
			break;

		case '+':
			newmode |= set->bits;
			break;

		case '-':
			newmode &= ~set->bits;
			break;

		case 'X':
			if (omode & (S_IFDIR|S_IXUSR|S_IXGRP|S_IXOTH))
				newmode |= set->bits;
			break;

		case '\0':
		default:
#ifdef SETMODE_DEBUG
			(void)printf("getmode:%04o -> %04o\n", omode, newmode);
#endif
			return (newmode);
		}
}

#define	ADDCMD(a, b, c, d)						\
	if (set >= endset) {						\
		BITCMD *newset;						\
		setlen += SET_LEN_INCR;					\
		newset = realloc(saveset, sizeof(BITCMD) * setlen);	\
		if (!newset) {						\
			if (saveset)					\
				free(saveset);				\
			saveset = NULL;					\
			return (NULL);					\
		}							\
		set = newset + (set - saveset);				\
		saveset = newset;					\
		endset = newset + (setlen - 2);				\
	}								\
	set = addcmd(set, (a), (b), (c), (d))

#define	STANDARD_BITS	(S_ISUID|S_ISGID|S_IRWXU|S_IRWXG|S_IRWXO)

void *
setmode(const char *p)
{
	int perm, who;
	char op, *ep;
	BITCMD *set, *saveset, *endset;
	sigset_t sigset, sigoset;
	mode_t mask;
	int equalopdone=0, permXbits, setlen;
	long perml;

	if (!*p)
		return (NULL);

	/*
	 * Get a copy of the mask for the permissions that are mask relative.
	 * Flip the bits, we want what's not set.  Since it's possible that
	 * the caller is opening files inside a signal handler, protect them
	 * as best we can.
	 */
	sigfillset(&sigset);
	(void)sigprocmask(SIG_BLOCK, &sigset, &sigoset);
	(void)umask(mask = umask(0));
	mask = ~mask;
	(void)sigprocmask(SIG_SETMASK, &sigoset, NULL);

	setlen = SET_LEN + 2;

	if ((set = malloc((u_int)(sizeof(BITCMD) * setlen))) == NULL)
		return (NULL);
	saveset = set;
	endset = set + (setlen - 2);

	/*
	 * If an absolute number, get it and return; disallow non-octal digits
	 * or illegal bits.
	 */
	if (isdigit((unsigned char)*p)) {
		perml = strtol(p, &ep, 8);
		if (*ep || perml < 0 || perml & ~(STANDARD_BITS|S_ISTXT)) {
			free(saveset);
			return (NULL);
		}
		perm = (mode_t)perml;
		ADDCMD('=', (STANDARD_BITS|S_ISTXT), perm, mask);
		set->cmd = 0;
		return (saveset);
	}

	/*
	 * Build list of structures to set/clear/copy bits as described by
	 * each clause of the symbolic mode.
	 */
	for (;;) {
		/* First, find out which bits might be modified. */
		for (who = 0;; ++p) {
			switch (*p) {
			case 'a':
				who |= STANDARD_BITS;
				break;
			case 'u':
				who |= S_ISUID|S_IRWXU;
				break;
			case 'g':
				who |= S_ISGID|S_IRWXG;
				break;
			case 'o':
				who |= S_IRWXO;
				break;
			default:
				goto getop;
			}
		}

getop:		if ((op = *p++) != '+' && op != '-' && op != '=') {
			free(saveset);
			return (NULL);
		}
		if (op == '=')
			equalopdone = 0;

		who &= ~S_ISTXT;
		for (perm = 0, permXbits = 0;; ++p) {
			switch (*p) {
			case 'r':
				perm |= S_IRUSR|S_IRGRP|S_IROTH;
				break;
			case 's':
				/* If only "other" bits ignore set-id. */
				if (!who || who & ~S_IRWXO)
					perm |= S_ISUID|S_ISGID;
				break;
			case 't':
				/* If only "other" bits ignore sticky. */
				if (!who || who & ~S_IRWXO) {
					who |= S_ISTXT;
					perm |= S_ISTXT;
				}
				break;
			case 'w':
				perm |= S_IWUSR|S_IWGRP|S_IWOTH;
				break;
			case 'X':
				permXbits = S_IXUSR|S_IXGRP|S_IXOTH;
				break;
			case 'x':
				perm |= S_IXUSR|S_IXGRP|S_IXOTH;
				break;
			case 'u':
			case 'g':
			case 'o':
				/*
				 * When ever we hit 'u', 'g', or 'o', we have
				 * to flush out any partial mode that we have,
				 * and then do the copying of the mode bits.
				 */
				if (perm) {
					ADDCMD(op, who, perm, mask);
					perm = 0;
				}
				if (op == '=')
					equalopdone = 1;
				if (op == '+' && permXbits) {
					ADDCMD('X', who, permXbits, mask);
					permXbits = 0;
				}
				ADDCMD(*p, who, op, mask);
				break;

			default:
				/*
				 * Add any permissions that we haven't already
				 * done.
				 */
				if (perm || (op == '=' && !equalopdone)) {
					if (op == '=')
						equalopdone = 1;
					ADDCMD(op, who, perm, mask);
					perm = 0;
				}
				if (permXbits) {
					ADDCMD('X', who, permXbits, mask);
					permXbits = 0;
				}
				goto apply;
			}
		}

apply:		if (!*p)
			break;
		if (*p != ',')
			goto getop;
		++p;
	}
	set->cmd = 0;
#ifdef SETMODE_DEBUG
	(void)printf("Before compress_mode()\n");
	dumpmode(saveset);
#endif
	compress_mode(saveset);
#ifdef SETMODE_DEBUG
	(void)printf("After compress_mode()\n");
	dumpmode(saveset);
#endif
	return (saveset);
}

static BITCMD *
addcmd(BITCMD *set, int op, int who, int oparg, u_int mask)
{
	switch (op) {
	case '=':
		set->cmd = '-';
		set->bits = who ? who : STANDARD_BITS;
		set++;

		op = '+';
		/* FALLTHROUGH */
	case '+':
	case '-':
	case 'X':
		set->cmd = op;
		set->bits = (who ? who : mask) & oparg;
		break;

	case 'u':
	case 'g':
	case 'o':
		set->cmd = op;
		if (who) {
			set->cmd2 = ((who & S_IRUSR) ? CMD2_UBITS : 0) |
				    ((who & S_IRGRP) ? CMD2_GBITS : 0) |
				    ((who & S_IROTH) ? CMD2_OBITS : 0);
			set->bits = (mode_t)~0;
		} else {
			set->cmd2 = CMD2_UBITS | CMD2_GBITS | CMD2_OBITS;
			set->bits = mask;
		}

		if (oparg == '+')
			set->cmd2 |= CMD2_SET;
		else if (oparg == '-')
			set->cmd2 |= CMD2_CLR;
		else if (oparg == '=')
			set->cmd2 |= CMD2_SET|CMD2_CLR;
		break;
	}
	return (set + 1);
}

#ifdef SETMODE_DEBUG
static void
dumpmode(BITCMD *set)
{
	for (; set->cmd; ++set)
		(void)printf("cmd: '%c' bits %04o%s%s%s%s%s%s\n",
		    set->cmd, set->bits, set->cmd2 ? " cmd2:" : "",
		    set->cmd2 & CMD2_CLR ? " CLR" : "",
		    set->cmd2 & CMD2_SET ? " SET" : "",
		    set->cmd2 & CMD2_UBITS ? " UBITS" : "",
		    set->cmd2 & CMD2_GBITS ? " GBITS" : "",
		    set->cmd2 & CMD2_OBITS ? " OBITS" : "");
}
#endif

/*
 * Given an array of bitcmd structures, compress by compacting consecutive
 * '+', '-' and 'X' commands into at most 3 commands, one of each.  The 'u',
 * 'g' and 'o' commands continue to be separate.  They could probably be
 * compacted, but it's not worth the effort.
 */
static void
compress_mode(BITCMD *set)
{
	BITCMD *nset;
	int setbits, clrbits, Xbits, op;

	for (nset = set;;) {
		/* Copy over any 'u', 'g' and 'o' commands. */
		while ((op = nset->cmd) != '+' && op != '-' && op != 'X') {
			*set++ = *nset++;
			if (!op)
				return;
		}

		for (setbits = clrbits = Xbits = 0;; nset++) {
			if ((op = nset->cmd) == '-') {
				clrbits |= nset->bits;
				setbits &= ~nset->bits;
				Xbits &= ~nset->bits;
			} else if (op == '+') {
				setbits |= nset->bits;
				clrbits &= ~nset->bits;
				Xbits &= ~nset->bits;
			} else if (op == 'X')
				Xbits |= nset->bits & ~setbits;
			else
				break;
		}
		if (clrbits) {
			set->cmd = '-';
			set->cmd2 = 0;
			set->bits = clrbits;
			set++;
		}
		if (setbits) {
			set->cmd = '+';
			set->cmd2 = 0;
			set->bits = setbits;
			set++;
		}
		if (Xbits) {
			set->cmd = 'X';
			set->cmd2 = 0;
			set->bits = Xbits;
			set++;
		}
	}
}
